content/ADR-138-social-service-authentication.md

```diff
@@ -0,0 +1,101 @@
+
+The service will have a distributed cache (without persistence) with user-to-token information, which will be used to obtain the user (once the token is validated) and to prevent hitting Matrix for every request.
+
+The login flow will be the following:
+
+```mermaid
+sequenceDiagram
+    User->>+Social service: Login
+    Social service->>+Matrix: Login
+    Matrix->>-Social service: Token
+    Social service->>+Cache: Store token
+    Social service-->>-User: Token
+```
+
+Then each subsequent request will perform the following validation:
+
+```mermaid
+graph LR
+    User -->|Request| Social-service(Social service)
+    Social-service -->|Get User | Cache
+    Cache -->|User| Social-service
+    Social-service --> C{User present}
+
+    C -->|True| D(Continue request)
+    C -->|False| Validate-token(Validate token against matrix)
+    Validate-token --> |Get User| Matrix
+    Matrix --> |User| Social-service
+```
+
+In this second workflow, the token life in the cache will be extended by what's defined in Matrix, thus preventing the user from using an invalid token.
+If the token is expired and the user tries to use it again, the service will validate against Matrix, and if it's still valid, the user will be obtained again and re-stored in the Cache.
+
+To validate the token against Matrix, an authenticated endpoint must be chosen. The endpoint needs to satisfy the following specifications:
+
+- No to low impact on the database
+- No to low usage from the client (to prevent overloading it even more)
+Ideally tackled by a specific worker, so in case it's necessary, it would be easier to scale up
+
+The endpoint that we have chosen is `_matrix/client/r0/pushrules/`, but could change if we find one that suits better.
+
+## Benefit
+
+This solution has two main benefits:
+
+1. There's no need to create a new authorization/authentication service or system
+1. The solution is completely seamless for the user and doesn't imply the persistence of tokens from the social service side
+
+## Competition
+
+<!--
+What other options were considered? Give an honest treatment of why
+these alternatives were not satisfactory. Identify the competition and
+demonstrate that the competition is clearly understood. Include the
+“what if we do nothing” alternative.
+-->
+
+### Implementing an authentication service
+
+This is a great of if we ever want to stop using Matrix. We would be in control of everything, allowing to build login features as we deem necessary. Although this brings a lot of flexibility on login features, the biggest problem of this option is the cost of implementing everything from scratch, which is not necessary due to the current requirements.
```

Comment:
Sorry, I made a mistake when I made the suggestion 🤦 
```suggestion
This is a great option if we ever want to stop using Matrix. We would be in control of everything, allowing to build login features as we deem necessary. Although this brings a lot of flexibility on login features, the biggest problem of this option is the cost of implementing everything from scratch, which is not necessary due to the current requirements.
```

Reply:
Yep, I already fixed it in a commit